### PR TITLE
Migrate genfrac, lap, smash, op, mod, and katex into their own files in src/functions

### LINKED ISSUE #869: migrate genfrac, lap, smash, op, mod, and katex into their own files in src/functions
This is a followup to #766 which did the same for phantom and delimsizing functions.

### PR DESCRIPTION
**Pull request by kevinbarabash**
Fixes #869

### CODE REVIEW

**Comment by ronkok** [comment]
Would it make sense to merge `op.js` and `operators.js` into one file? That was my long-term intent when I wrote the PR for `\operatorname`.

**Comment by kevinbarabash** [comment]
@ronkok looking at `op.js` and `operators.js` there isn't really any overlap.  I'm going to rename `operators.js` to `operatorname.js` to follow the pattern of each file in `src/functions` being responsible for the group type it defines.

**Comment by ronkok** [comment]
@kevinbarabash I agree completely. I didn't know how thin you wanted to slice the new files. If I had, I would have it named it that way myself.

File: src/buildHTML.js [hunk 1]
```diff
@@ -446,141 +446,6 @@ groupTypes.supsub = function(group, options) {
         options);
 };
 
-groupTypes.genfrac = function(group, options) {
```
Comment (kevinbarabash) [inline src/buildHTML.js]: moved to `htmlBuilder` in `genfrac.js`.

File: src/buildHTML.js [hunk 2]
```diff
@@ -603,305 +468,7 @@ groupTypes.spacing = function(group, options) {
     }
 };
 
-groupTypes.lap = function(group, options) {
```
Comment (kevinbarabash) [inline src/buildHTML.js]: moved to `htmlBuilder` in `lap.js`.

File: src/buildHTML.js [hunk 3]
```diff
@@ -603,305 +468,7 @@ groupTypes.spacing = function(group, options) {
...
-    } else {
-        inner = makeSpan(
-            ["inner"], [buildGroup(group.value.body, options)]);
-    }
-    const fix = makeSpan(["fix"], []);
-    return makeSpan(
-        ["mord", group.value.alignment], [inner, fix], options);
-};
-
-groupTypes.smash = function(group, options) {
```
Comment (kevinbarabash) [inline src/buildHTML.js]: moved to `htmlBuilder` in `smash.js`.

File: src/buildHTML.js [hunk 4]
```diff
@@ -603,305 +468,7 @@ groupTypes.spacing = function(group, options) {
...
-    // But the span still has an HTML line height.
-    // makeVList applies "display: table-cell", which prevents the browser
-    // from acting on that line height. So we'll call makeVList now.
-
-    return buildCommon.makeVList([
-        {type: "elem", elem: node},
-    ], "firstBaseline", null, options);
-};
-
-groupTypes.op = function(group, options) {
```
Comment (kevinbarabash) [inline src/buildHTML.js]: moved to `htmlBuilder` in `op.js`.

File: src/buildHTML.js [hunk 5]
```diff
@@ -603,305 +468,7 @@ groupTypes.spacing = function(group, options) {
...
-        if (baseShift) {
-            base.style.position = "relative";
-            base.style.top = baseShift + "em";
-        }
-
-        return base;
-    }
-};
-
-groupTypes.mod = function(group, options) {
```
Comment (kevinbarabash) [inline src/buildHTML.js]: moved to `htmlBuilder` in `mod.js`.

File: src/buildHTML.js [hunk 6]
```diff
@@ -603,305 +468,7 @@ groupTypes.spacing = function(group, options) {
...
-    }
-
-    if (group.value.modType === "pod" || group.value.modType === "pmod") {
-        inner.push(buildCommon.mathsym(")", group.mode));
-    }
-
-    return buildCommon.makeFragment(inner);
-};
-
-groupTypes.katex = function(group, options) {
```
Comment (kevinbarabash) [inline src/buildHTML.js]: moved to `htmlBuilder` in `katex.js`.

File: src/buildMathML.js [hunk 7]
```diff
@@ -247,49 +247,6 @@ groupTypes.supsub = function(group, options) {
     return node;
 };
 
-groupTypes.genfrac = function(group, options) {
```
Comment (kevinbarabash) [inline src/buildMathML.js]: moved to `mathmlBuilder` in `genfrac.js`.

File: src/buildMathML.js [hunk 8]
```diff
@@ -341,66 +298,6 @@ groupTypes.spacing = function(group) {
     return node;
 };
 
-groupTypes.op = function(group, options) {
```
Comment (kevinbarabash) [inline src/buildMathML.js]: move to `mathmlBuilder` in `op.js`.

File: src/buildMathML.js [hunk 9]
```diff
@@ -341,66 +298,6 @@ groupTypes.spacing = function(group) {
...
-            "mi", [new mathMLTree.TextNode(group.value.body.slice(1))]);
-
-        // TODO(ron): Append an <mo>&ApplyFunction;</mo> as in \operatorname
-        // ref: https://www.w3.org/TR/REC-MathML/chap3_2.html#sec3.2.2
-    }
-
-    return node;
-};
-
-groupTypes.mod = function(group, options) {
```
Comment (kevinbarabash) [inline src/buildMathML.js]: moved to `mathmlBuilder` in `mod.js`.

File: src/buildMathML.js [hunk 10]
```diff
@@ -341,66 +298,6 @@ groupTypes.spacing = function(group) {
...
-    }
-    if (group.value.modType === "pod" || group.value.modType === "pmod") {
-        inner.push(new mathMLTree.MathNode(
-            "mo", [makeText(")", group.mode)]));
-    }
-
-    return new mathMLTree.MathNode("mo", inner);
-};
-
-groupTypes.katex = function(group) {
```
Comment (kevinbarabash) [inline src/buildMathML.js]: moved to `mathmlBuilder` in `katex.js`

File: src/buildMathML.js [hunk 11]
```diff
@@ -571,35 +468,6 @@ groupTypes.kern = function(group) {
     return node;
 };
 
-groupTypes.lap = function(group, options) {
```
Comment (kevinbarabash) [inline src/buildMathML.js]: moved to `mathmlBuilder` in `lap.js`.

File: src/buildMathML.js [hunk 12]
```diff
@@ -571,35 +468,6 @@ groupTypes.kern = function(group) {
...
-    if (group.value.alignment !== "rlap")    {
-        const offset = (group.value.alignment === "llap" ? "-1" : "-0.5");
-        node.setAttribute("lspace", offset + "width");
-    }
-    node.setAttribute("width", "0px");
-
-    return node;
-};
-
-groupTypes.smash = function(group, options) {
```
Comment (kevinbarabash) [inline src/buildMathML.js]: moved to `mathmlBuilder` in `smash.js`.

File: src/functions.js [hunk 13]
```diff
@@ -336,147 +309,15 @@ defineFunction([
...
-}, function(context, args) {
-    const body = args[0];
-    return {
-        type: "op",
-        limits: false,
-        symbol: false,
-        value: ordargument(body),
-    };
-});
+import "./functions/operatorname";
```
Comment (kevinbarabash) [inline src/functions.js]: changed from `import "./functions/operators";` on line 368 before the diff.

File: src/functions.js [hunk 14]
```diff
@@ -336,147 +309,15 @@ defineFunction([
...
 
-import "./functions/operators";
+import "./functions/genfrac";
 
-// Fractions
-defineFunction([
-    "\\dfrac", "\\frac", "\\tfrac",
-    "\\dbinom", "\\binom", "\\tbinom",
-    "\\\\atopfrac", // can’t be entered directly
```
Comment (kevinbarabash) [inline src/functions.js]: moved to `genfrac.js` and refactored to use the new `defineFunction`.

File: src/functions.js [hunk 15]
```diff
@@ -336,147 +309,15 @@ defineFunction([
...
-        limits: true,
-        symbol: true,
-        body: context.funcName,
-    };
-});
+import "./functions/op";
 
-// \mathop class command
-defineFunction(["\\mathop"], {
-    numArgs: 1,
```
Comment (kevinbarabash) [inline src/functions.js]: moved to `op.js` and refactored to use the new `defineFunction` (second call).

File: src/functions.js [hunk 16]
```diff
@@ -336,147 +309,15 @@ defineFunction([
     };
 });
 
-// Limits, symbols
-defineFunction([
-    "\\coprod", "\\bigvee", "\\bigwedge", "\\biguplus", "\\bigcap",
-    "\\bigcup", "\\intop", "\\prod", "\\sum", "\\bigotimes",
-    "\\bigoplus", "\\bigodot", "\\bigsqcup", "\\smallint",
```
Comment (kevinbarabash) [inline src/functions.js]: moved to `op.js` and refactored to use the new `defineFunction` (first call).

File: src/functions.js [hunk 17]
```diff
@@ -229,27 +222,7 @@ defineFunction(["\\stackrel"], {
...
-    numArgs: 0,
-}, function(context, args) {
-    return {
-        type: "mod",
-        modType: "bmod",
-        value: null,
-    };
-});
-
-defineFunction(["\\pod", "\\pmod", "\\mod"], {
```
Comment (kevinbarabash) [inline src/functions.js]: moved to `mod.js` and refactored to use the new `defineFunction` (second call).

File: src/functions.js [hunk 18]
```diff
@@ -229,27 +222,7 @@ defineFunction(["\\stackrel"], {
     };
 });
 
-// \mod-type functions
-defineFunction(["\\bmod"], {
```
Comment (kevinbarabash) [inline src/functions.js]: moved to `mod.js` and refactored to use the new `defineFunction` (first call).

File: src/functions.js [hunk 19]
```diff
@@ -175,14 +175,7 @@ defineFunction(["\\kern", "\\mkern"], {
     };
 });
 
-// A KaTeX logo
-defineFunction(["\\KaTeX"], {
```
Comment (kevinbarabash) [inline src/functions.js]: moved to `katex.js` and refactored to use the new `defineFunction`.

File: src/functions.js [hunk 20]
```diff
@@ -336,147 +309,15 @@ defineFunction([
...
-        hasBarLine: hasBarLine,
-        leftDelim: leftDelim,
-        rightDelim: rightDelim,
-        size: size,
-    };
-});
+import "./functions/lap";
 
-// Horizontal overlap functions
-defineFunction(["\\mathllap", "\\mathrlap", "\\mathclap"], {
```
Comment (kevinbarabash) [inline src/functions.js]: moved to `lap.js` and refactored to use the new `defineFunction`.

File: src/functions.js [hunk 21]
```diff
@@ -336,147 +309,15 @@ defineFunction([
...
-    const body = args[0];
-    return {
-        type: "lap",
-        alignment: context.funcName.slice(5),
-        body: body,
-    };
-});
-
-// smash, with optional [tb], as in AMS
-defineFunction(["\\smash"], {
```
Comment (kevinbarabash) [inline src/functions.js]: moved to `smash.js` and refactored to use the new `defineFunction`.

**Comment by marcianx** [comment]
Man, is there _any_ way whatsoever to get new files to diff against specific old file from which the code came? I'm having to do manual diffing for each new file against the pre-modified `functions.js` and it's cumbersome.

**Comment by kevinbarabash** [comment]
In retrospect I shouldn't have tried to do so many at the same time.  If you want, I can split this up into separate PRs.  TBH, the diff is just a bunch of copy/pasting with minor changes so splitting this diff wouldn't be hard.

**Comment by kevinbarabash** [comment]
I'm going to close this and will re-open as separate, smaller PRs.

**Comment by marcianx** [comment]
@kevinbarabash IMO, even leaving it as a large PR would have been fine for me if it were a sequence of small commits that are independently easy to review.

**Comment by edemaine** [comment]
@marcianx Cool! FWIW, you could use for loops to help automate the process:
```
for a in buildHTML.js buildMathML.js functions.js
do
  for b in genfrac.js katex.js lap.js mod.js op.js smash.js
  do
    git difftool HEAD~:src/$a src/functions/$b
  done
done
```

**Comment by kevinbarabash** [comment]
@edemaine will the code you posted run as a shell script?

**Comment by edemaine** [comment]
@kevinbarabash Yes, or you can type it directly into your shell, assuming you're using bash (or another sh clone) as your shell.

**Comment by kevinbarabash** [comment]
@marcianx let me know if you're having second thoughts about reviewing this as is.  I'm willing and able to split this into smaller PRs.

**Comment by marcianx** [comment]
Sorry. Really busy weekend and minor accident right before. Will try to look at this soon.

**Comment by kevinbarabash** [comment]
@marcianx hope you're alright.  If you do find that you don't have time please let me know.

**Comment by marcianx** [comment]
- [ ] genfrac.js
- [ ] katex.js
- [ ] lap.js
- [ ] mod.js
- [ ] op.js
- [ ] smash.js
- [ ] sanity-check rest of the changes

**Comment by marcianx** [comment]
Nice! For posterity, this is how I reviewed. I went per file as in the checklist above and wanted to open all the diffs for a given file in the same meld session. So I temporarily added the `-n` parameter to my `.gitconfig` to open all diffs in the same session:

    [difftool "meld"]
        cmd = meld -n \"$LOCAL\" \"$REMOTE\" --label \"DIFF (ORIGINAL MY)\"

and per file, I did the for loop per @edemaine's suggestion modified to open all diffs at once (e.g. for `smash.js`):

    for a in buildHTML.js buildMathML.js functions.js; do
      git difftool HEAD~:src/$a src/functions/smash.js &
    done

and did some easy manual modifications to get clean diffs. It was all pretty fast for this cut-and-paste-and-indent PR.

**Comment by marcianx** [comment]
@kevinbarabash I'll let you submit after rebasing.

**Comment by kevinbarabash** [comment]
@marcianx thanks so much for reviewing this and also sharing your method.  I'll rebase sometime tomorrow.

**Pull request merged**